src/lib/blkdata_svc/blk_read_tracker.cpp

```diff
@@ -28,59 +28,39 @@ void BlkReadTracker::merge(const BlkId& blkid, int64_t new_ref_count,
                            const std::shared_ptr< blk_track_waiter >& waiter) {
     HS_DBG_ASSERT(new_ref_count ? waiter == nullptr : waiter != nullptr, "Invalid waiter");
 
-    //
-    // Don't move alignment handling outside of this function, because the nblks between (first and last blk num after
-    // alignment) could be larger than 255 which exceeds a BlkId can hold;
-    //
-    auto cur_blk_num_aligned = s_cast< blk_num_t >(sisl::round_down(blkid.blk_num(), entries_per_record()));
-    auto last_blk_num_aligned_up =
-        s_cast< blk_num_t >(sisl::round_up(blkid.blk_num() + blkid.blk_count() + 1, entries_per_record()) - 1);
+    auto cur_base_blk_num = s_cast< blk_num_t >(sisl::round_down(blkid.blk_num(), entries_per_record()));
+    auto last_base_blk_num =
+        s_cast< blk_num_t >(sisl::round_down(blkid.blk_num() + blkid.blk_count() - 1, entries_per_record()));
 
     [[maybe_unused]] bool waiter_rescheduled{false};
     // everything is aligned after this point, so we don't need to handle sub_range in a base blkid;
-    while (cur_blk_num_aligned <= last_blk_num_aligned_up) {
-        BlkId base_blkid{cur_blk_num_aligned, entries_per_record(), blkid.chunk_num()};
-
-        BlkTrackRecord rec;
-        const auto rec_found = m_pending_reads_map.get(base_blkid, rec);
-
-        if (new_ref_count != 0) {
-            // this is insert/remove operations
-            if (rec_found) {
-                // if some read is already happening on this record, just update the ref_cnt;
+    while (cur_base_blk_num <= last_base_blk_num) {
+        BlkId base_blkid{cur_base_blk_num, entries_per_record(), blkid.chunk_num()};
+
+        if (new_ref_count > 0) {
+            m_pending_reads_map.upsert_or_delete(base_blkid,
+                                                 [&base_blkid, new_ref_count](BlkTrackRecord& rec, bool existing) {
+                                                     if (!existing) { rec.m_key = base_blkid; }
+                                                     rec.m_ref_cnt += new_ref_count;
```

Comment:
nit: I knwn `m_ref_cnt` is defaulted to zero, would suggest set the ref_cnt if not exist for readability. 
```if (!existing) { 
    rec.m_key = base_blkid;
    rec.m_ref_cnt = new_ref_cnt;
}  else { rec.m_ref_cnt += new_ref_cnt; }```